### PR TITLE
edit the deleting logic of dataset and alluxioruntime

### PR DESCRIPTION
**Pull request by yangyuliufeng**
### Ⅰ. Describe what this PR does
This PR edits the deleting logic of dataset and alluxioruntime , so that delete dataset can cascading delete runtime. 
The alluxioruntime needs to query the cache through the dataset when it is deleted. Now , it can be deleted without a dataset.
During the deleting logic, I have found a bug : when deleting alluxioruntime, the pvc can be deleted if the pvc name is not same as the volume name defined in pod.Spec.Volumes
So , I changed the calibration condition from  "if pvc.Name == pvcName" in function GetPvcMountPods to "pvc.PersistentVolumeClaim.ClaimName == pvcName".

### Ⅱ. Does this pull request fix one issue?

#405
#420
### Ⅳ. Describe how to verify it
when a pod is using the pvc ,delete the dataset, it cannot be deleted.
When deleting a dataset , the alluxioruntime will be deleted later.
### Ⅴ. Special notes for reviews
the return values in cache.go are not consistent， I change them to  "states".

### CODE REVIEW

File: pkg/controllers/runtime_controller.go [hunk 1]
```diff
@@ -119,6 +121,7 @@ func (r *RuntimeReconciler) ReconcileInternal(ctx cruntime.ReconcileRequestConte
 		return result, err
 	}
 
+	// 6. Add Finalizer of runtime
```
Comment (cheyang) [inline pkg/controllers/runtime_controller.go:124]: Add finalizer是否可以再提前一点？

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/ddc/alluxio/cache.go [hunk 2]
```diff
@@ -53,28 +42,30 @@ func (e *AlluxioEngine) queryCacheStatus() (states cacheStates, err error) {
...
 			// Convert Alluxio's binary byte units to Fluid's binary byte units
 			// e.g. 10KB -> 10KiB, 2GB -> 2GiB
-			usedCacheCapacity = utils.BytesSize(float64(usedCacheCapacityAlluxio))
+			states.cached = utils.BytesSize(float64(usedCacheCapacityAlluxio))
 		}
 	}
 
+	dataset, err := utils.GetDataset(e.Client, e.name, e.namespace)
+	if err != nil {
+		e.Log.Info("Failed to get dataset when query cache status")
```
Comment (TrafalgarZZZ) [inline pkg/ddc/alluxio/cache.go:57]: 此处是否对NotFound Error和其他Error分开处理更直观？例如使用`utils.IgnoreNotFound`将NotFound Error忽略掉，`cleanUpCache()`方法中也可以进而修改为对`err != nil`的判断

**Review by TrafalgarZZZ — CHANGES_REQUESTED** [review]

File: pkg/controllers/runtime_controller.go (hunk 1 above)
Comment (xieydd) [reply pkg/controllers/runtime_controller.go:124]: 正确的逻辑是？：
1. 获取 runtime
2. Add Finalizer of runtime
3. Get the dataset 
4. Get the ObjectMeta of runtime
5. Reconcile delete the runtime
6. update status of dataset

File: pkg/controllers/runtime_controller.go [hunk 3]
```diff
@@ -263,6 +266,26 @@ func (r *RuntimeReconciler) AddFinalizerAndRequeue(ctx cruntime.ReconcileRequest
 	return utils.RequeueImmediatelyUnlessGenerationChanged(prevGeneration, currentGeneration)
 }
 
+// AddFinalizerAndRequeue add  finalizer and requeue
```
Comment (xieydd) [inline pkg/controllers/runtime_controller.go:269]: AddOwnerAndRequeue add OwnerReferences for runtime metadata and requeue

File: pkg/controllers/runtime_controller.go (hunk 1 above)
Comment (yangyuliufeng) [reply pkg/controllers/runtime_controller.go:124]: @xieydd 我觉得应该是
1、获取runtime
2、ReconcileRuntimeDeletion（这个可以放到前面来）
3、Get the ObjectMeta of runtime
4、Get the dataset
5、if dataset exist：
（1）Add Finalizer of runtime（这个需要在Get the ObjectMeta of runtime之后）
（2）AddOwnerAndRequeue
7、update status of dataset

File: pkg/ddc/alluxio/cache.go (hunk 2 above)
Comment (yangyuliufeng) [reply pkg/ddc/alluxio/cache.go:57]: @TrafalgarZZZ 由于这个方法不止是删除的时候调用，改了err的返回逻辑，还会影响其它方法的判定

Comment (TrafalgarZZZ) [reply pkg/ddc/alluxio/cache.go:57]: @yangyuliufeng OK，同意。我只是认为这里有一个很小的Corner Case：如果`GetDataset`返回了一个不是NotFound的Error，那么`cleanupCache`端目前是不会去Check它的。

**Pull request closed without merging**